### PR TITLE
fix(jest): improving spyOn for better auto complete suggestions

### PR DESCRIPTION
**Pull request by Farenheith**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/facebook/jest/blob/24e0472ac41ea03cdd89ffaea8c5f410ecf6054f/packages/jest-jasmine2/src/jasmine/spyRegistry.ts>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### Additional info about the change:

The current spyOn overload declarations provide a 100% fit over all the use cases we have, but they don't allow a good enough code complete suggestion. This change doesn't really add anything new to the signatures, but it improves code complete suggestions.

For example, take a look at how the suggestions are shown before the change:
image

And how it is shown after:
image

The return type still matches the previous one, the only real difference is that, when doing something like that:

```ts
jest.spyOn(console, 'error', undefined);
```

The error will be a bit different.
Before the change:
```
No overload matches this call.
  Overload 1 of 4, '(object: Console, method: "Console", accessType: "get"): SpyInstance<ConsoleConstructor, []>', gave the following error.
    Argument of type '"error"' is not assignable to parameter of type '"Console"'.
  Overload 2 of 4, '(object: Console, method: "Console", accessType: "set"): SpyInstance<void, [ConsoleConstructor]>', gave the following error.
    Argument of type '"error"' is not assignable to parameter of type '"Console"'.ts(2769)
```
After the change:
```
Argument of type 'undefined' is not assignable to parameter of type 'never'.ts(2345)
```

@sandersn @ErfanMirzapour @G-Rath FYK

### CODE REVIEW

**Review by G-Rath — APPROVED** [review]
Can confirm in WebStorm the autosuggest seems "stronger" (it would already show the right values as it would resolve them itself but now it's showing the actual properties as it's being provided the info from TypeScript which is more fulsome)

**Comment by GerkinDev** [comment]
LGTM

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -282,16 +282,24 @@ declare namespace jest {
...
+        accessType: A,
+    ): M extends NonFunctionPropertyNames<Required<T>>
+        ? A extends 'set'
+            ? SpyInstance<void, [Required<T>[M]]>
+            : SpyInstance<Required<T>[M], []>
+        : Required<T>[M] extends new (...args: any[]) => any
+        ? SpyInstance<InstanceType<Required<T>[M]>, ConstructorArgsType<Required<T>[M]>>
+        : Required<T>[M] extends (...args: any) => any
+        ? SpyInstance<ReturnType<Required<T>[M]>, ArgsType<Required<T>[M]>>
+        : never;
```
Comment (GerkinDev) [inline types/jest/index.d.ts:302]: Ooff this is rough. Wouldn't it be possible to comment it a bit or make it more maintainable ? Splitting in other generic conditional types ? I don't know if it matches code policies, but it is almost impossible to review.

Comment (Farenheith) [reply types/jest/index.d.ts:302]: Sure, I'll play around a little to see what I can come up with

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -282,16 +282,21 @@ declare namespace jest {
...
-    ): SpyInstance<void, [Required<T>[M]]>;
+        method: Key,
+        accessType: A,
+    ):
+      A extends SetAccessor ? SpyInstance<void, [Value]>
+    : A extends GetAccessor ? SpyInstance<Value, []>
+    : Value extends Constructor ? SpyInstance<InstanceType<Value>, ConstructorArgsType<Value>>
+    : Value extends Func ? SpyInstance<ReturnType<Value>, ArgsType<Value>>
+    : never;
     function spyOn<T extends {}, M extends FunctionPropertyNames<Required<T>>>(
```
Comment (Farenheith) [inline types/jest/index.d.ts:300]: @GerkinDev can you take another look? I refactored the return time to make it more readable, creating some new auxiliary types and letting the sequence of conditions a bit clearer.

Comment (GerkinDev) [reply types/jest/index.d.ts:300]: Way better, thank you :)

**Review by GerkinDev — APPROVED** [review]

**Pull request merged**